DOCS-553: Refactor Motion Service page (#1116)
viamrobotics/docs: docs/services/motion/algorithms.md

> ---

Author: Is this page important? @spieswl ? It feels to me like it's geared towards technical users who are curious, but doesn't really matter for someone who just wants to use the Motion Service.

Constraints are linked to from the Move method section where relevant, but yeah I can link to both from the intro if you think these pages are worth calling more attention to?

> ---

Reviewer: This page is highly technical and all about implementation details, so if we already have a standard practice to make this kind of content prominent, I have no reservations about that. Otherwise, only the most technically savvy will get in here, so I don't believe it is important to highlight.

title: "About the Motion Planning Algorithms"
linkTitle: "Algorithms"
weight: 20
type: "docs"
description: "Information about the motion planning algorithms Viam uses."
---

> Viam implements two planning algorithms, both based in principle on [RRT](https://en.wikipedia.org/wiki/Rapidly-exploring_random_tree):

Reviewer 2: I think before diving into each algorithm, it would be good to explain here that Viam first starts with one and then uses the other (the last paragraph on this page currently). And then dive into each algorithm

Author: I added a heading down there, but I'm not sure how to move it up here when the concepts haven't been introduced yet.

Author: ```suggestion
Robotic motion planning is heavily reliant on planning algorithms to determine how to achieve motion for a particular scenario.
Many algorithms already exist for this problem, and motion planning is a domain where improvements and novel developments occur frequently.
Viam does not implement all motion planning algorithms but has implemented two in its strategy for planning general robot motion.
The two algorithms Viam supports are both based in principle on [RRT](https://en.wikipedia.org/wiki/Rapidly-exploring_random_tree):
```

## RRT*-Connect

RRT*-Connect is an asymptotically optimal planner that samples the planning space randomly, connecting viable paths as it finds them.
It continues sampling after it finds its first valid path, and if it finds future paths that are more efficient, it updates to report those instead.
For Viam, efficiency/path quality is measured in terms of total kinematics state excursion.

> For an arm, this refers to joints; the total amount of joint change is minimized.

Reviewer: ```suggestion
For an arm, the total sum of the arms joint changes is minimized.

```

Author: ```suggestion
For an arm, the sum of all joint changes is minimized.
```

> For a gantry, this refers to the amount of linear movement.

Author: ```suggestion
For a gantry, the sum of all linear movement is minimized.
```
Is this okay? and gosh darn GitHub with the sneaky newline bugs :(

> For a gantry, this refers to the amount of linear movement.

Reviewer: Yep this one is good too.

This algorithm can route around obstacles, but cannot satisfy topological constraints.

## CBiRRT

CBiRRT stands for Constrained, Bi-directional implementation of [RRT](https://en.wikipedia.org/wiki/Rapidly-exploring_random_tree).
It creates paths which conform to specified constraints, and attempts to smooth them afterwards as needed.
By default, it does not constrain the path of motion at all.
This is to ensure that paths will be found when using defaults.
CBiRRT returns the first valid path that it finds.

> The CBiRRT algorithm Viam uses is based on the algorithm described in this paper: <https://www.ri.cmu.edu/pub_files/2009/5/berenson_dmitry_2009_2.pdf>

Reviewer 2: ```suggestion
The CBiRRT algorithm Viam uses is based on the algorithm described in the paper [Manipulation Planning on Constraint Manifolds](https://www.ri.cmu.edu/pub_files/2009/5/berenson_dmitry_2009_2.pdf)

```

---

> By default, Viam's motion planning library uses a hybrid approach.

Reviewer 2: Can I change this? If so how?

Author: @spieswl I have no idea

Reviewer: You technically can, but we've specifically tried to avoid exposing this to the user so far. There are extras options which allow for more control over Motion Planning operation, but the `planning manager` that handles the hybrid approach was designed so that users would not need to get so deep into the technical details.

Author: So should I delete the "By default"?
